packages/logging/lokiLogProvider.go

```diff
@@ -0,0 +1,274 @@
+	return &LokiLogProvider{
+		gatewayUrl: gatewayUrl,
+		username:   username,
+		password:   password,
+		clock:      &realClock{},
+	}
+}
+
+func (llp *LokiLogProvider) getLogs(doguName string, linesCount int) ([]logLine, error) {
+	query := fmt.Sprintf("{pod=~\"%s.*\"}", doguName)
+	endDate := llp.clock.Now()
+	startDate := createQueryStartDateFromEndDate(endDate)
+
+	result := make([]logLine, 0)
+	for {
+		limit := calculateQueryLimit(linesCount, len(result))
+
+		logrus.Debugf("running loki query for '%s' from %s to %s with limit %d", doguName, startDate.Format(time.RFC3339), endDate.Format(time.RFC3339), limit)
+		lokiQueryUrl, err := buildLokiQueryUrl(llp.gatewayUrl, query, startDate, endDate, limit)
+		if err != nil {
+			return result, fmt.Errorf("failed to build loki-query: %v", err)
+		}
+
+		lokiResponse, err := llp.doLokiHttpQuery(lokiQueryUrl)
+		if err != nil {
+			return result, fmt.Errorf("faild to execute loki-query: %v", err)
+		}
+
+		logLines, err := extractLogLinesFromLokiResponse(lokiResponse)
+		if err != nil {
+			return nil, fmt.Errorf("failed to extract logs from loki response: %v", err)
+		}
+
+		if len(logLines) <= 0 {
+			// no new lines to read => nothing is left
+			break
+		}
+
+		// prepend logs
+		result = append(logLines, result...)
+
+		if len(logLines) < limit {
+			// the query returned fewer lines than requested => nothing is left
+			break
+		}
+
+		if linesCount > 0 && len(result) >= linesCount {
+			// we reached the maximum lines count
+			break
+		}
+
+		// there are still logs to read -> start with the newest log timestamp from the last response
+		endDate = logLines[0].timestamp
+		startDate = createQueryStartDateFromEndDate(endDate)
+	}
+
+	// because multiple logs can happen at the exact same timestamp and the query is batched over time,
+	// it is possible that consecutive batches contain the same logLines. These need to be removed.
+	result = deduplicateLogLines(result)
+
+	// truncate log-lines if we got more than requested
+	if linesCount > 0 && len(result) > linesCount {
+		result = result[:linesCount]
+	}
+
+	logrus.Debugf("finished loki query; got %d logLines", len(result))
+
+	return result, nil
+}
+
+func calculateQueryLimit(linesCount int, resultCount int) int {
+	if linesCount <= 0 {
+		return defaultQueryLimit
+	}
+
+	remainingCount := linesCount - resultCount
+
+	if remainingCount < 0 {
+		return 0
+	}
+
+	if (remainingCount) < defaultQueryLimit {
+		return remainingCount
+	}
+	return defaultQueryLimit
+}
+
+// createQueryStartDateFromEndDate calculates the start date for a loki query based on the given end date.
+// Since loki query run backwards (in time) the start date is calculated based on the end date.
+// The start date is set 30 days before the given end date, because that is the maximum range that loki allows.
+func createQueryStartDateFromEndDate(endDate time.Time) time.Time {
+	return endDate.Add(-24 * 30 * time.Hour)
+}
```

Comment:
Do we really want to query the maximum here? Cause that could be a lot of logs, most of which we'll never need.

Reply:
The limit is always set to the linecount given by the user or to 1000 (when the given linecount is 0)
So there should never be more than 1000 lines in the result.

Reply:
I mean the line count in the request, which will be 30 days of logs if I'm understanding correctly. That will likely be a lot of network overhead.

Reply:
Ah sorry, I understand now